File: tests/test_memory.py
"""Cold source loading must fit a capped allocator, including parser temporaries."""
from pathlib import Path
import os
import subprocess
import sys
import tempfile
import unittest

ROOT = Path(__file__).resolve().parents[1]


class MemoryBudget(unittest.TestCase):
    def test_cold_source_load(self):
        runner = ROOT / ('.build/lua-memory.exe' if os.name == 'nt' else '.build/lua-memory')
        compiler = ROOT / ('.build/luac53.exe' if os.name == 'nt' else '.build/luac53')
        if not runner.exists():
            subprocess.run([sys.executable, 'tools/memory_test.py'], cwd=ROOT, check=True)
        self.assertTrue(compiler.exists(), 'run tools/build53.py first')
        modules = ('params', 'wire', 'pview', 'pdb', 'pinput')
        with tempfile.TemporaryDirectory() as cache:
            for name in modules:
                subprocess.run([str(compiler), '-s', '-o', str(Path(cache) / (name + '.luac')),
                                str(ROOT / 'src/SCRIPTS/MAV' / (name + '.lua'))], check=True)
            # Opening the Parameters page must stay close to the level the radio already runs.
            # The index builder is loaded on demand, so it does not inflate the cold-open peak:
            # v0.1.3 needed 136 KiB and this build needs about 140, which is the budget below.
            # A regression that loads the builder eagerly, or grows these modules, breaks it.
            for history, limit in (('', 141 * 1024), ('history', 147 * 1024)):
                env = dict(os.environ, MAV_TEST_HEAP_LIMIT=str(limit))
                result = subprocess.run([str(runner), 'tests/test_memory.lua', 'src', history, cache],
                                        cwd=ROOT, env=env, capture_output=True, text=True)
                self.assertEqual(result.returncode, 0, result.stdout + result.stderr)

    def test_first_open_compiles_once_then_uses_cache(self):
        """A package now ships a .luac per module, so an installed open must never compile.

        The first open of a package built without caches does compile each module, and that
        compile is the peak a pilot pays once. Every later open must load bytecode instead:
        recompiling while a cache exists re-pays the whole peak and is what made the radio
        report 'not enough memory' on an open which then worked on retry.
        """
        runner = ROOT / ('.build/lua-memory.exe' if os.name == 'nt' else '.build/lua-memory')
        compiler = ROOT / ('.build/luac53.exe' if os.name == 'nt' else '.build/luac53')
        if not runner.exists():
            subprocess.run([sys.executable, 'tools/memory_test.py'], cwd=ROOT, check=True)
        with tempfile.TemporaryDirectory() as cache:
            for name in ('params', 'wire', 'pview', 'pdb', 'pinput'):
                subprocess.run([str(compiler), '-s', '-o', str(Path(cache) / (name + '.luac')),
                                str(ROOT / 'src/SCRIPTS/MAV' / (name + '.lua'))], check=True)
            for mode, expected in (('nocache', 'First open: compiled'), ('', 'Warm open: loaded entirely from cache')):
                env = dict(os.environ, MAV_TEST_HEAP_LIMIT=str(152 * 1024))
                argv = [str(runner), 'tests/test_memory.lua', 'src', '', cache, '-', mode] if mode else \
                       [str(runner), 'tests/test_memory.lua', 'src', '', cache]
                result = subprocess.run(argv, cwd=ROOT, env=env, capture_output=True, text=True)
                self.assertEqual(result.returncode, 0, result.stdout + result.stderr)
                self.assertIn(expected, result.stdout)

    def test_builder_loads_within_budget(self):
        """Requesting a build loads the index builder, which must also fit a capped allocator."""
        runner = ROOT / ('.build/lua-memory.exe' if os.name == 'nt' else '.build/lua-memory')
        compiler = ROOT / ('.build/luac53.exe' if os.name == 'nt' else '.build/luac53')
        if not runner.exists():
            subprocess.run([sys.executable, 'tools/memory_test.py'], cwd=ROOT, check=True)
        with tempfile.TemporaryDirectory() as cache:
            for name in ('params', 'wire', 'pview', 'pdb', 'pinput', 'index'):
                subprocess.run([str(compiler), '-s', '-o', str(Path(cache) / (name + '.luac')),
                                str(ROOT / 'src/SCRIPTS/MAV' / (name + '.lua'))], check=True)
            env = dict(os.environ, MAV_TEST_HEAP_LIMIT=str(163 * 1024))
            result = subprocess.run([str(runner), 'tests/test_memory.lua', 'src', '', cache, 'builder'],
                                    cwd=ROOT, env=env, capture_output=True, text=True)
            self.assertEqual(result.returncode, 0, result.stdout + result.stderr)
            self.assertIn('Builder bytes live/peak:', result.stdout)

            # Prove the shipped cache is what makes a build affordable. With the opening set
            # cached but the builder itself not, the build must COMPILE it, which is far larger
            # than the budget above. That is the failure pilots hit: a compile which exhausts
            # the heap never writes a cache, so the build failed identically on every retry.
            # Asserting the failure keeps the reason for shipping .luac caches honest, and would
            # fail if the builder ever shrank enough to make the cache unnecessary.
            result = subprocess.run([str(runner), 'tests/test_memory.lua', 'src', '', cache,
                                     'builder', 'buildfirst'], cwd=ROOT, env=env,
                                    capture_output=True, text=True)
            self.assertNotEqual(result.returncode, 0,
                                'uncached builder compile unexpectedly fits the budget')
            # The harness returns nil when EdgeTX's loader cannot compile, so the assertion is
            # what trips rather than an explicit allocation message.
            self.assertIn('test_memory.lua:103', result.stdout + result.stderr)


if __name__ == '__main__':
    unittest.main()

File: tests/test_memory.lua
-- Tiny host harness: measure cold source compilation with the core retained.
-- The allocator counter also captures temporary parser allocations between calls.
local root = arg[1] or 'src'
local cacheRoot = assert(arg[3], 'stripped cache root required')
LCD_W, LCD_H, SOLID, INVERS = 128, 96, 0, 1
EVT_PAGE_FIRST, EVT_PAGE_BREAK, EVT_VIRTUAL_ENTER, EVT_VIRTUAL_NEXT = 40, 41, 42, 43
local screen, remaining = '', 0
getTime = function() return 100 end
getRSSI = function() return 0 end
getValue = function() return 0 end
getFieldInfo = function() return nil end
killEvents = function() end
crossfireTelemetryPop = function()
  if remaining == 0 then return end
  local message = string.rep('W', 48) .. string.format('%02d', remaining)
  remaining = remaining - 1
  local p = {0xF1, 4}
  for i=1,#message do p[#p+1]=message:byte(i) end
  return 0x80, p
end
local sent = 0
crossfireTelemetryPush = function() sent = sent + 1 return true end
io.seek = function() error('opening Parameters must not access cache') end
io.open = function() return true end
io.close = function() end
lcd = {clear=function() screen='' end, drawLine=function() end,
  drawText=function(_,y,s) if y<LCD_H then screen=screen..s..'\n' end end}
-- Model EdgeTX's loader honestly so the cached and first-open paths are both covered.
-- Both packages now ship a .luac beside each .lua, so an installed open loads bytecode. A
-- package built without caches has none, so its first open compiles each module and the
-- firmware then writes a cache; later opens load bytecode directly. Forcing compilation while
-- a valid cache exists re-pays the whole compile peak, which is the regression this guards
-- against, so it is counted and asserted rather than tolerated.
--
-- 'buildfirst' models the real radio sequence: the page already opened successfully, so the
-- five opening modules are cached, but a build has never run and so its module is not. That
-- is the case which exhausted memory on hardware before the builder's cache shipped, and it is
-- invisible to the other modes.
local mode = arg[5]
local warm = mode ~= 'nocache' and mode ~= 'buildfirst'
local written, compiles = {}, 0
if mode == 'buildfirst' then
  for _, name in ipairs({'wire', 'pview', 'pdb', 'pinput', 'params'}) do
    written[name .. '.luac'] = true
  end
end
loadScript = function(path, mode)
  local name = path:match('[^/]+$')
  if path:sub(-5) == '.luac' then
    if not (warm or written[name]) then return nil end
    return loadfile(cacheRoot .. '/' .. name)
  end
  assert(path:sub(-4) == '.lua', 'module source path must be explicit')
  assert(mode == 'tc' or mode == 'tx', 'unexpected source load mode ' .. tostring(mode))
  if mode == 'tc' then
    compiles = compiles + 1
    written[name:gsub('%.lua$', '.luac')] = true
  end
  return loadfile(root .. path)
end
collectgarbage('collect')
memoryStats(true)
local app = assert(loadfile(root .. '/SCRIPTS/TELEMETRY/MAV.lua'))()
app.init() app.run(0)
if arg[2] == 'history' then
  remaining = 20
  for i=1,20 do app.run(0) collectgarbage('step', 10) end
end
collectgarbage('collect')
print('Navigation bytes live/peak:', memoryStats())
app.run(40) app.run(40)
for i=1,12 do app.run(0) collectgarbage('step', 10) end
assert(screen:find('> LOAD <', 1, true), screen)
assert(sent == 0, 'opening Parameters must not transmit')
print('Parameters bytes live/peak:', memoryStats())
app.run(42) app.run(0)
assert(screen:find('Connecting...', 1, true), screen)
assert(sent == 1, 'Load starts one bounded bridge subscription')
print('Load bytes live/peak:', memoryStats())
-- A warm open must be cache-only. Compiling while a cache exists re-pays the whole compile
-- peak and is what made the radio report "not enough memory" on an open that then worked on
-- retry. A first open has no cache and legitimately compiles each module once; the build-first
-- case has the opening modules cached and must compile only the builder.
if mode == 'buildfirst' then
  assert(compiles == 0, 'build-first open compiled ' .. compiles .. ' already-cached module(s)')
elseif warm then
  assert(compiles == 0, 'warm open recompiled ' .. compiles .. ' cached module(s)')
else
  assert(compiles >= 5, 'first open must compile the module set once, saw ' .. compiles)
end
print(mode == 'buildfirst' and 'Build-first open: opening set all cached' or
  warm and 'Warm open: loaded entirely from cache' or
  ('First open: compiled ' .. compiles .. ' module(s)'))
-- The index builder is loaded on demand, only when a build is requested, so its cost lands
-- on the build path rather than on every open. This measures that incremental cost, which is
-- the figure that matters for a radio: a build must fit alongside the already-open app.
-- Loaded the way loadParameterModule does: prefer the cache, compile only when absent.
if arg[4] == 'builder' then
  local before = memoryStats(true)
  -- Mirrors loadParameterModule: cached bytecode when present, otherwise compile once.
  local chunk = loadScript('/SCRIPTS/MAV/index.luac', 'b')
  if not chunk then chunk = loadScript('/SCRIPTS/MAV/index.lua', 'tc') end
  local maker = assert(chunk)()
  collectgarbage('collect')
  local after = memoryStats()
  assert(type(maker) == 'function', 'index builder module loads and returns a factory')
  print('Builder bytes live/peak:', after, '(delta from Load:', after - before, ')')
end
